graphql/batch_executor.go

```diff
@@ -0,0 +1,409 @@
+	for idx, source := range sources {
+		slice := reflect.ValueOf(source)
+		respList := make([]interface{}, slice.Len())
+		for i := 0; i < slice.Len(); i++ {
+			writer := newObjectWriter(destinations[idx], strconv.Itoa(idx)) // TODO Parent?
+			respList[i] = writer
+			flattenedResps = append(flattenedResps, writer)
+			flattenedSources = append(flattenedSources, slice.Index(i).Interface())
+		}
+		destinations[idx].Fill(respList)
+	}
+	return unwrapBatchResult(ctx, flattenedSources, typ.Type, selectionSet, flattenedResps)
+}
+
+// Traverses the Union type and resolves or creates exec units to resolve
+// all of the sub-objects for all the provided sources.
+func unwrapBatchUnionResult(ctx context.Context, sources []interface{}, typ *Union, selectionSet *SelectionSet, destinations []*objectWriter) ([]*ExecutionUnit, error) {
+	sourcesByType := make(map[string][]interface{}, len(typ.Types))
+	destinationsByType := make(map[string][]*objectWriter, len(typ.Types))
+	for idx, src := range sources {
+		union := reflect.ValueOf(src)
+		if union.Kind() == reflect.Ptr && union.IsNil() {
+			// Don't create a destination for any nil Unions types
+			destinations[idx].Fill(nil)
+			continue
+		}
+
+		srcType := ""
+		if union.Kind() == reflect.Ptr && union.Elem().Kind() == reflect.Struct {
+			union = union.Elem()
+		}
+		for typString := range typ.Types {
+			inner := union.FieldByName(typString)
+			if inner.IsNil() {
+				continue
+			}
+			if srcType != "" {
+				return nil, fmt.Errorf("union type field should only return one value, but received: %s %s", srcType, typString)
+			}
+			srcType = typString
+			sourcesByType[srcType] = append(sourcesByType[srcType], inner.Interface())
+			destinationsByType[srcType] = append(destinationsByType[srcType], destinations[idx])
+		}
+	}
+
+	var execUnits []*ExecutionUnit
+	for srcType, sources := range sourcesByType {
+		gqlType := typ.Types[srcType]
+		for _, fragment := range selectionSet.Fragments {
+			if fragment.On != srcType {
+				continue
+			}
+			units, err := unwrapBatchObjectResult(ctx, sources, gqlType, fragment.SelectionSet, destinationsByType[srcType])
+			if err != nil {
+				return nil, err
+			}
+			execUnits = append(execUnits, units...)
+		}
+
+	}
+	return execUnits, nil
+}
+
+// Traverses the object selections and resolves or creates exec units to resolve
+// all of the object fields for every source passed in.
+func unwrapBatchObjectResult(ctx context.Context, sources []interface{}, typ *Object, selectionSet *SelectionSet, destinations []*objectWriter) ([]*ExecutionUnit, error) {
+	selections := Flatten(selectionSet)
+	numExpensive := 0
+	numNonExpensive := 0
+	for _, selection := range selections {
+		if field, ok := typ.Fields[selection.Name]; ok && field.Expensive {
+			numExpensive++
+		} else if ok && field.External {
+			numNonExpensive++
+		}
+	}
+
+	// For every object, create a "destination" map that we can fill with our
+	// result values.  Filter out invalid/nil objects.
```

Comment:
why don't slices have to do this?

Reply:
I think we get away with it because slices become zero value sized.